common/locks/lock_test.go

```diff
@@ -95,6 +96,34 @@ func BenchmarkParallelStdlibLock(b *testing.B) {
 	})
 }
 
+func TestShouldBeFatal(t *testing.T) {
+	t.Run("stdlib", func(t *testing.T) {
+		fatalTest(t)
+		var m sync.Mutex
+		m.Unlock()
+	})
+	t.Run("custom", func(t *testing.T) {
+		fatalTest(t)
+		NewMutex().Unlock()
+	})
+}
+
+func fatalTest(t *testing.T) {
+	t.Helper()
+	if os.Getenv("RUN_FATAL_TESTS") != "true" {
+		t.Skipf(
+			"This test intentionally crashes the process, and cannot be tested normally.\n"+
+				"Check output manually when running this test on its own:\n"+
+				"\tRUN_FATAL_TESTS=true go test -test.run %q %v",
+			t.Name(), "github.com/uber/cadence/common/locks",
+		)
+	} else {
+		t.Cleanup(func() {
+			t.Fatal("process should have crashed, and this should be unreachable")
+		})
+	}
+}
```

Comment:
We should probably consider making some sort of standard pattern for tests like this, to make sure they correctly-fail in CI after e.g. Go upgrades.

Go's stdlib accomplishes this by running `go [stuff]` as an `exec.Command` in tests, and... that'd work, but I think it's the sort of thing we should write exactly once, and then everything should use it, rather than yoloing.  It seems rather high-risk for CI complications.